### PR TITLE
Add Locals, Variables, Resources, and Outputs

### PR DESCRIPTION
**Pull request by luke-taylor**
## Summary 
Initial configuration has been designed to create both a VPN Gateway and ExpressRoute Gateway, however the module has been additionally configured to connect a Local Network Gateway to the VPN Gateway. This will be extended for the analogous Express Route Circuits.

I have tried to artificially separate VPN arguments and ExpressRoute arguments for (hopefully) a better end-user experience, the tedious part is deciphering the VPN arguments from the ExpressRoute ones in the `azurerm_virtual_network_gateway` resource, the ones that are strictly vpn related I have prefixed with `vpn_`.  If it's required on both it's prefixed `gateway_`. Let me know what you think. 

Question:
1. Do we want to bound all the resources to one region? At the moment I have opened a configuration argument to change the location of the LGW and Connection if needed, I am not sure of the utility of this however. 

Here is a snippet of what calling this module could look like: 

```hcl 

resource "azurerm_resource_group" "rg" {
  name     = "rg-connectivity-uksouth-prod"
  location = "uksouth"
}

resource "azurerm_virtual_network" "vnet" {
  name                = "vnet-uksouth-prod"
  resource_group_name = azurerm_resource_group.rg.name
  location            = azurerm_resource_group.rg.location
  address_space       = ["10.0.0.0/16"]
}

module "vgw" {
  source = "../.."

  location                      = "uksouth"
  resource_group_name           = azurerm_resource_group.rg.name
  virtual_network_name          = azurerm_virtual_network.vnet.name
 
  gateway_name                  = "vgw-uksouth-prod"
  gateway_subnet_address_prefix = "10.0.0.1/24"
  gateway_type                  = "Vpn"
  gateway_sku                   = "VpnGw1"
  gateway_ip_configurations = [
    {
      ip_configuration_name        = null
      apipa_addresses              = null
      private_ip_allocation_method = null
      public_ip = {
        name              = "pip-vgw-001"
        allocation_method = null
        sku               = null
        tags              = null
      }
    }
  ]

  vpn_bgp_enabled = true
  vpn_bgp_settings = {
    asn         = 65515
    peer_weight = 0
  }

  local_network_gateways = [
    {
      name       = "lgw-uksouth-prod"
      location   = "uksouth"
      ip_address = "x.x.x.x"
      connection_config = {
        name           = "lgw-uksouth-prod-to-vgw-uksouth-prod"
        type           = "IPsec"
        routing_weight = 0
      }
    }
  ]
}


```

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -1,3 +1,175 @@
-resource "null_resource" "nop" {
+resource "azurerm_subnet" "vgw" {
```
Comment (jaredfholgate) [inline main.tf:1]: Should / Could there be a BYO subnet option?

File: main.tf [hunk 2]
```diff
@@ -1,3 +1,175 @@
...
+  }
+
+  depends_on = [
+    azurerm_subnet.vgw,
+    azurerm_public_ip.vgw
+  ]
+}
+
+resource "azurerm_local_network_gateway" "vgw" {
+  for_each = length(var.local_network_gateways) == 0 ? {} : { for local_network_gateway in var.local_network_gateways : local_network_gateway.name => local_network_gateway }
```
Comment (jaredfholgate) [inline main.tf:69]: I think the ternary may be unnecessary here? The variable defaults to an empty array.

File: main.tf [hunk 3]
```diff
@@ -1,3 +1,175 @@
...
+  service_provider_name    = each.value.service_provider_name
+  peering_location         = each.value.peering_location
+  bandwidth_in_mbps        = each.value.bandwidth_in_mbps
+  bandwidth_in_gbps        = each.value.bandwidth_in_gbps
+  allow_classic_operations = each.value.allow_classic_operations
+  express_route_port_id    = each.value.express_route_port_id
+  authorization_key        = each.value.authorization_key
+  tags                     = each.value.tags
+
+  dynamic "sku" {
```
Comment (jaredfholgate) [inline main.tf:106]: The sku block is Required, so can't be an empty array.

File: main.tf [hunk 4]
```diff
@@ -1,3 +1,175 @@
...
+    content {
+      asn                 = each.value.bgp_settings.asn
+      bgp_peering_address = each.value.bgp_settings.bgp_peering_address
+      peer_weight         = each.value.bgp_settings.peer_weight
+    }
+  }
+}
+
+resource "azurerm_express_route_circuit" "er" {
+  for_each = length(var.express_route_circuits) == 0 ? {} : { for express_route_circuit in var.express_route_circuits : express_route_circuit.name => express_route_circuit }
```
Comment (jaredfholgate) [inline main.tf:91]: I think this ternary may be unnecessary? The local should produce an empty array.

File: main.tf [hunk 5]
```diff
@@ -1,3 +1,175 @@
...
+    content {
+      tier   = each.value.sku.tier
+      family = each.value.sku.family
+    }
+  }
+
+}
+
+resource "azurerm_virtual_network_gateway_connection" "lgw" {
+  for_each = length(local.virtual_network_gateway_connections) == 0 ? {} : { for virtual_network_gateway_connection in local.virtual_network_gateway_connections : virtual_network_gateway_connection.name => virtual_network_gateway_connection }
```
Comment (jaredfholgate) [inline main.tf:118]: I think this ternary may be uneccesary?

File: variables.tf [hunk 6]
```diff
@@ -1,17 +1,215 @@
...
+  type        = string
+}
+
+variable "gateway_edge_zone" {
+  description = "The availability zone of the virtual network gateway. Only supported for AZ SKUs."
+  type        = string
+  default     = null
+}
+
+variable "gateway_type" {
```
Comment (jaredfholgate) [inline variables.tf:39]: Add validation to this variable if it only ever allows two values.

File: variables.tf [hunk 7]
```diff
@@ -1,17 +1,215 @@
...
+  description = "The type of the virtual network gateway, ExpressRoute or VPN."
+  type        = string
+}
+
+variable "gateway_sku" {
```
Comment (jaredfholgate) [inline variables.tf:44]: Could validation be added here?

File: variables.tf [hunk 8]
```diff
@@ -1,17 +1,215 @@
...
+  description = "The SKU (size) of the virtual network gateway."
+  type        = string
+}
+
+variable "vpn_generation" {
```
Comment (jaredfholgate) [inline variables.tf:49]: Validation?

File: variables.tf [hunk 9]
```diff
@@ -1,17 +1,215 @@
...
+variable "vpn_bgp_enabled" {
+  description = "Enable BGP for the virtual network gateway."
+  type        = bool
+  default     = false
+}
+
+variable "vpn_private_ip_address_enabled" {
+  description = "Enable private IP address for the virtual network gateway for Virtual Network Gateway Connections. Only supported for AZ SKUs."
+  type        = bool
+  default     = null
```
Comment (jaredfholgate) [inline variables.tf:82]: Should this be false rather than null?

**Review by jaredfholgate — COMMENTED** [review]
Hi Luke. This module looks great. My comments are really minor things that you were probably going to clean up anyway. 

In terms of the variable names, I think they are fine as they are. I like to use root level variables for simplicity whenever I can rather than nested objects. The only real alternative to what you have chosen would be nested objects, so a `vpn_settings` and `express_route_settings` top level variable with the specifics as properties on those objects. I am not sure this would make the module easier to use though, but think either approach is fine.

More that happy to have a call later this week if you think I can help at all.

File: main.tf (hunk 1 above)
Comment (luke-taylor) [reply main.tf:1]: Yes there could, perhaps a `subnet_creation_enabled` variable defaulted to true?

File: main.tf (hunk 2 above)
Comment (luke-taylor) [reply main.tf]: Yes apologies for some of the sloppy logic, I think most of these conditionals are gone now after the latest push.

File: main.tf (hunk 4 above)
Comment (luke-taylor) [reply main.tf]: Yes apologies for some of the sloppy logic, I think most of these conditionals are gone now after the latest push. 

Also, it's been decided that this module shouldn't be creating the ExpressRoute Circuits themselves only the configure post provisioning.

File: main.tf (hunk 5 above)
Comment (luke-taylor) [reply main.tf]: Yes apologies for some of the sloppy logic, I think most of these conditionals are gone now after the latest push.

File: variables.tf (hunk 6 above)
Comment (luke-taylor) [reply variables.tf]: Good suggestion, Jared. Added.

**Comment by luke-taylor** [comment]
> Hi Luke. This module looks great. My comments are really minor things that you were probably going to clean up anyway.
> 
> In terms of the variable names, I think they are fine as they are. I like to use root level variables for simplicity whenever I can rather than nested objects. The only real alternative to what you have chosen would be nested objects, so a `vpn_settings` and `express_route_settings` top level variable with the specifics as properties on those objects. I am not sure this would make the module easier to use though, but think either approach is fine.
> 
> More that happy to have a call later this week if you think I can help at all.

Hey @jaredfholgate thanks for the feedback! I agree, I was trying to go with the policy of using root level variables on the resources that are only created strictly once (e.g. the gateway). I thought about an ExpressRoute Settings block for the gateway however the is no real strict configuration arguments for the ExpressRoute Gateway, a lot of the configuration (BGP etc.) for ExpressRoute happens on the Express Circuit Circuit/ Virtual Network Gateway Connection directly. 

What will be available though on my latest push, is a variable called `express_route_circuits` which contains an object called `peering_config` where a user can setup BGP etc. for a pre-existing ExpressRoute Circuit. 

Also, I apologize about the messiness of the module, some of the logic isn't making sense at the moment like you've pointed out I am hoping this latest push has sorted that out. 

I'm OOF until the 11th of July so I'll try get some time with you after then and go through it in more detail.

**Pull request merged**